src/Microsoft.AspNet.Mvc.Razor/DefaultRazorPageFactory.cs

```diff
@@ -58,18 +58,20 @@ public IRazorPage CreateInstance(string relativePath)
                 // For tilde slash paths, drop the leading ~ to make it work with the underlying IFileProvider.
                 relativePath = relativePath.Substring(1);
             }
-
             var result = CompilerCache.GetOrAdd(relativePath, _compileDelegate);
-
-            if (result == CompilerCacheResult.FileNotFound)
+            if (result.IsFoundResult)
             {
-                return null;
+                return new RazorPageFactoryResult(() =>
+                {
+                    var page = (IRazorPage)Activator.CreateInstance(result.CompilationResult.CompiledType);
+                    page.Path = relativePath;
+                    return page;
+                }, result.ExpirationTokens);
```

Comment:
What's up with this wonky lambda thing? Why doesn't this class just take the required parameters in the constructor?